Support Kodi Style Movie Extras

antgiant opened the pull request:
This pull request adds support for scanning the Extras folder. It utilizes the same logic as scanning inside the root folder for extras with the addition of untagged files in the extras folder are given a default tag.

Both Kodi and Emby require that all Movie Extras be stored in a subfolder titled Extras ([Kodi Requirements](http://kodi.wiki/view/Add-on:VideoExtras#Extras_Folder)  [Emby Requirements](https://github.com/MediaBrowser/Wiki/wiki/Movie%20naming#user-content-movie-extras)). Plex treats the extras folder [exactly the same](https://github.com/plexinc-plugins/Scanners.bundle/blob/d8ca7d014d07ff68f3013a2728801a5ad50a6325/Contents/Resources/Common/VideoFiles.py#L14) as the folder structures it does support, but it does not scan inside it. 

Accepting this pull request will allow users of both Kodi and Emby to access load their movies without having to rearrange their file system. ([Here](https://forums.plex.tv/discussion/192704/extras-folder-naming-compatible-with-both-plex-and-kodi) is someone requesting exactly that.)

mmccurdy commented:
This code is no longer maintained in this repo, but have a look at this diff, which I think does what you're asking for:

* Consider `-deletedscene` the same as `-deleted`
* Add all videos directly in "Extras" folders according to their `-type`, defaulting to `other`

What bug are you trying to fix specifically in a111e85?

```
diff --git a/Bundles/LocalMedia.bundle/Contents/Code/localmedia.py b/Bundles/LocalMedia.bundle/Contents/Code/localmedia.py
index 786cb78..ba4f200 100644
--- a/Bundles/LocalMedia.bundle/Contents/Code/localmedia.py
+++ b/Bundles/LocalMedia.bundle/Contents/Code/localmedia.py
@@ -86,7 +86,7 @@ def findAssets(metadata, media_title, paths, type, parts=[]):
 
       # Don't count things that follow the "-extra" naming convention.
       if should_count and find_extras:
-        for key in extra_type_map.keys():
+        for key in extra_type_map.keys() + ['deletedscene']:
           if root.endswith('-' + key):
             Log('%s looks like a %s extra, won\'t contribute to total media file count.' % (file_path, key))
             should_count = False
@@ -129,21 +129,31 @@ def findAssets(metadata, media_title, paths, type, parts=[]):
         Log('Looking for local extras in path: ' + path)
         for root, dirs, files in os.walk(path):
           for d in dirs:
-            for key in extra_type_map.keys():
-              if re_strip.sub('', d.lower()).startswith(key):
-                for f in os.listdir(os.path.join(root, d)):
-                  (fn, ext) = os.path.splitext(f)
-                  if not fn.startswith('.') and ext[1:] in config.VIDEO_EXTS:
-
-                    # On Windows, os.walk() likes to prepend the "extended-length path prefix" to root.
-                    # This causes issues later on when this path is converted to the file:// URL for
-                    # serialization and later consumption by PMS, so clean it up here.
-                    #
-                    root = re.sub(r'^\\\\\?\\', '', root)
-                    
-                    Log('Found %s extra: %s' % (key, f))
-                    extras.append({'type' : key, 'title' : helpers.unicodize(fn), 'file' : os.path.join(root, d, f)})
-                continue
+
+            # We want to add all videos directly in directories named exactly "Extras," giving them a default type if necessary.
+            if d.lower() == 'extras':
+              Log('Adding all extras from directory: %s' % os.path.join(root, d))
+              for f in os.listdir(os.path.join(root, d)):
+                (fn, ext) = os.path.splitext(f)
+                if not fn.startswith('.') and ext[1:] in config.VIDEO_EXTS:
+                  extras.append({'type' : extra_type_map.get(fn.split('-')[-1], 'other'), 'title' : helpers.unicodize(fn), 'file' : os.path.join(re.sub(r'^\\\\\?\\', '', root), d, f)})
+
+            else:
+              for key in extra_type_map.keys():
+                if re_strip.sub('', d.lower()).startswith(key):
+                  for f in os.listdir(os.path.join(root, d)):
+                    (fn, ext) = os.path.splitext(f)
+                    if not fn.startswith('.') and ext[1:] in config.VIDEO_EXTS:
+
+                      # On Windows, os.walk() likes to prepend the "extended-length path prefix" to root.
+                      # This causes issues later on when this path is converted to the file:// URL for
+                      # serialization and later consumption by PMS, so clean it up here.
+                      #
+                      root = re.sub(r'^\\\\\?\\', '', root)
+
+                      Log('Found %s extra: %s' % (key, f))
+                      extras.append({'type' : key, 'title' : helpers.unicodize(fn), 'file' : os.path.join(root, d, f)})
+                  continue
 
         # Look for filenames following the "-extra" convention and a couple of other special cases.
         for f in os.listdir(path):
@@ -158,7 +168,11 @@ def findAssets(metadata, media_title, paths, type, parts=[]):
           # Files following the "-extra" convention.
           else:
             for key in extra_type_map.keys():
-              if not fn.startswith('.') and fn.endswith('-' + key) and ext[1:] in config.VIDEO_EXTS:
+              fn_key = key
+              if key == 'deleted' and fn.endswith('-deletedscene'):
+                Log('Treating file as a "deleted" extra using alternate naming: %s' % fn)
+                fn_key = 'deletedscene'
+              if not fn.startswith('.') and fn.endswith('-' + fn_key) and ext[1:] in config.VIDEO_EXTS:
                 Log('Found %s extra: %s' % (key, f))
                 title = ' '.join(fn.split('-')[:-1])
                 extras.append({'type' : key, 'title' : helpers.unicodize(title), 'file' : os.path.join(path, f)})
```

antgiant commented:
That looks great! Thank you! 

The bug in a111e853503ded72a414ebfb730baa6810cd2c57 was simply that when I constructed the path the the extras I accidentally left out "Extras" resulting in paths like Movie/extra-scene.mpg instead of the correct Movie/Extras/extra-scene.mpg. Your code correctly handles that.

antgiant commented:
@mmccurdy I just applied your patch to my test environment and found a few problems. In particular this line has two problems.
```
+                  extras.append({'type' : extra_type_map.get(fn.split('-')[-1], 'other'), 'title' : helpers.unicodize(fn), 'file' : os.path.join(re.sub(r'^\\\\\?\\', '', root), d, f)})
````
First it doesn't have the logic to catch `-deletedscene` tagged files. Second, much more critically extra_type_map.get() returns a full object instead of just the key string. As a result the entire extra scan breaks and always returns no extras.

Here is a patch to be applied on top of your patch that fixes these issues.
```
diff --git a/Contents/Code/localmedia.py b/Contents/Code/localmedia.py
index ba4f200..eceb2cc 100644
--- a/Contents/Code/localmedia.py
+++ b/Contents/Code/localmedia.py
@@ -132,11 +132,24 @@ def findAssets(metadata, media_title, paths, type, parts=[]):
 
             # We want to add all videos directly in directories named exactly "Extras," giving them a default type if necessary.
             if d.lower() == 'extras':
-              Log('Adding all extras from directory: %s' % os.path.join(root, d))
+              Log('Adding all extras from directory: %s' % re.sub(r'^\\\\\?\\', '', os.path.join(root, d)))
               for f in os.listdir(os.path.join(root, d)):
                 (fn, ext) = os.path.splitext(f)
-                if not fn.startswith('.') and ext[1:] in config.VIDEO_EXTS:
-                  extras.append({'type' : extra_type_map.get(fn.split('-')[-1], 'other'), 'title' : helpers.unicodize(fn), 'file' : os.path.join(re.sub(r'^\\\\\?\\', '', root), d, f)})
+                found = False
+                for key in extra_type_map.keys():
+                  fn_key = key
+                  if key == 'deleted' and fn.endswith('-deletedscene'):
+                    Log('Treating file as a "deleted" extra using alternate naming: %s' % fn)
+                    fn_key = 'deletedscene'
+                  if not fn.startswith('.') and fn.endswith('-' + fn_key) and ext[1:] in config.VIDEO_EXTS:
+                    Log('Found %s extra: %s' % (key, f))
+                    title = ' '.join(fn.split('-')[:-1])
+                    extras.append({'type' : key, 'title' : helpers.unicodize(title), 'file' : os.path.join(path, d, f)})
+                    found = True
+                if not found:
+                  Log('Found %s extra: %s' % ('other', f))
+                  extras.append({'type' : 'other', 'title' : helpers.unicodize(fn), 'file' : os.path.join(path, d, f)})
+                continue
 
             else:
               for key in extra_type_map.keys():
```

antgiant commented:
@mmccurdy It appears this still hasn't been rolled into the master. Anything I can do to help make that possible?

ScroogeHD117 commented:
Could this please be merged?

antgiant commented:
Considering the previous reaction I gave up on Plex being willing to accept patches. However, if that has changed I’d be willing to update this patch if it needs it.
